### DISCUSSION TITLE
how to type beancount.core.data.* items in kwargs?

### DISCUSSION THREAD

**Discussion by airtonix** [beancount/beancount#842] [Q&A]
Looking at how I'm supposed to use types and the objects out of `beancount.core.data`

Starting off with:

```py
class Transformer[T]:

  def transform(self, transaction: T, value: str) -> T:

    transaction.postings[1].account = value;
```

expected error: 
![image](https://github.com/user-attachments/assets/d8a3bd51-449a-49b0-885d-334aadeb0953)

Then I want to make sure that `T` is constrained to something like a `beancount.core.data.Transaction`; I run into issues (probably skill issues).

```py
TTransaction = TypeVar("TTransaction",bound=beancount_datatypes.Transaction)

class Transformer[TTransaction]:

  def transform(self, transaction: TTransaction, value: str) -> TTransaction:

    transaction.postings[1].account = value;
``` 

![image](https://github.com/user-attachments/assets/14a273e5-412f-4419-9186-089447da0936)

uhh. ok skill issue.

But what about: 

```py
TTransaction = beancount_datatypes.Transaction

class Transformer[TTransaction]:

  def transform(self, transaction: TTransaction, value: str) -> TTransaction:

    transaction.postings[1].account = value

    return transaction
```

nope. 

https://github.com/user-attachments/assets/2ee7e505-799d-4b6d-b03d-4eda5ba20d54

I'm familiar with Typescript, but here in python I'm a bit lost.

**Comment by trim21** [comment]
```

class Transformer:

  def transform(self, transaction: data.Transaction, value: str) -> data.Transaction:

    transaction.postings[1].account = value;
```
